File: AutoIt.Foundation/AutoIt.Foundation.Web.Scripts/ClassHelper/StringHelper.ts
interface String{
    NextPoint(startPoint: LinePoint,count: number): LinePoint;
    PrePoint(startPoint: LinePoint,count: number): LinePoint;
    MatchNext(regex: string, index?: number);
    MatchPre(regex: string, index?: number);
    Repeat(count: number): string;
    Reverse(): string;

    PrePoint1(startPoint: LinePoint, count: number): LinePoint;
}

//获取下一个位置（开始位置,偏移数）
String.prototype.NextPoint = function (startPoint,count) {
    var x = startPoint.X;
    var y = startPoint.Y;

    //从开始位置向后偏移指定字符数(从当前字符开始计算)
    for (var i = startPoint.Index; i < startPoint.Index + count; i++) {
        //移到下一行
        if (this[i] == '\n') {
            x = 0;
            y += 1;

        }
        //忽略回车符
        else if (this[i] == '\r') {

        }
        //移到下一列
        else {
            x += 1;
        }
    }

    var endPoint = new LinePoint(startPoint.Index + count, x, y);

    return endPoint;
}

//获取上一个位置(开始位置,偏移数)
String.prototype.PrePoint = function (startPoint,count) {
    var x = startPoint.X;
    var y = startPoint.Y;

    //从开始位置向前偏移指定字符数(从上一个字符开始计算)
    for (var i = startPoint.Index - 1; i > startPoint.Index - 1 - count; i--) {
         //移到上一行
        if (this[i] == '\n') {
            //获取上一行的内容
            var val = this.MatchPre('[^\\n]+', i - 1);
            x = val.length;
            y -= 1;
        }
        //忽略回车符
        else if (this[i] == '\r') {

        }
        //移到上一列
        else {
            x -= 1;
        }
    }

    var endPoint = new LinePoint(startPoint.Index - count, x, y);

    return endPoint;
}

//获取下一个匹配(匹配正则,匹配开始位置)
String.prototype.MatchNext = function(regex, index) {
    index = index || 0;

    if (index >= this.length) {
        return null;
    }

    var val = this.substr(index);
    var match = new RegExp(regex, "gm").exec(val);
    var result = match ? match[0] : "";

    return result;
}

//获取上一个匹配(匹配正则,匹配开始位置)
String.prototype.MatchPre = function (regex, index) {
    index = index || 0;

    if (index < 0) {
        return null;
    }

    //反转内容
    var val = this.substr(0, index + 1).Reverse();
    var match = new RegExp(regex, "g").exec(val);
    var result = match ? match[0] : "";
    //反转结果
    result = result.Reverse();

    return result;
}

//构建重复字符串(重复次数)
String.prototype.Repeat=function(count) {
    var val = "";

    for (var i = 0; i < count; i++) {
        val += this;
    }

    return val;
}

//反转字符串
String.prototype.Reverse=function() {
    return this.split('').reverse().join('');
}

File: AutoIt.Foundation/AutoIt.Foundation.Common/Basic/ClassHelper/StringHelper.cs
using System.Text.RegularExpressions;

namespace AutoIt.Foundation.Common
{
    public static class StringHelper
    {
        /// <summary>
        /// 获取下一个位置
        /// </summary>
        /// <param name="startPoint">开始位置</param>
        /// <param name="count">跳过的元素个数</param>
        public static LinePoint NextPoint(this string str, LinePoint startPoint,int count)
        {
            var x = startPoint.X;
            var y = startPoint.Y;

            for (var i = startPoint.Index; i < startPoint.Index+count; i++)
            {
                //换行
                if (str[i] == '\n')
                {
                    x = 0;
                    y += 1;
                    
                }
                else if (str[i] == '\r')
                {
                   
                }
                else
                {
                    x += 1;
                }
            }

            var endPoint=new LinePoint(startPoint.Index+count,x,y);

            return endPoint;
        }

        /// <summary>
        /// 获取前一个位置
        /// </summary>
        /// <param name="startPoint">开始位置</param>
        /// <param name="count">跳过的元素个数</param>
        public static LinePoint PrePoint(this string str, LinePoint startPoint, int count)
        {
            var x = startPoint.X;
            var y = startPoint.Y;

            for (var i = startPoint.Index; i > startPoint.Index - count; i--)
            {
                //换行 x=0???
                if (str[i] == '\n')
                {
                    x = -1;
                    y -= 1;

                    var tmpIndex = i - 1;
                    while (tmpIndex >= 0 && str[tmpIndex] != '\n')
                    {
                        if (str[tmpIndex] != '\r')
                        {
                            x++;
                        }

                        tmpIndex--;
                    }
                }
                else if (str[i] == '\r')
                {

                }
                else
                {
                    x -= 1;
                }
            }

            var endPoint = new LinePoint(startPoint.Index - count, x, y);

            return endPoint;
        }

        /// <summary>
        /// 获取下一个匹配
        /// </summary>
        public static string MatchNext(this string str, string regex, int? index)
        {
            if (index == null)
            {
                index = 0;
            }

            if (index >= str.Length)
            {
                return null;
            }

            var val = str.Substring(index.Value);
            var result = Regex.Match(val, regex, RegexOptions.IgnoreCase | RegexOptions.Multiline).Value;

            return result;
        }

        /// <summary>
        /// 获取上一个匹配(RegexOptions.RightToLeft)
        /// </summary>
        public static string MatchPre(this string str, string regex, int? index)
        {
            if (index ==null)
            {
                index = str.Length - 1;
            }

            if (index < 0)
            {
                return null;
            }

            var val = str.Substring(0, index.Value + 1);
            var result =
                Regex.Match(val, regex, RegexOptions.IgnoreCase | RegexOptions.Multiline | RegexOptions.RightToLeft)
                    .Value;

            return result;
        }
    }
}
